File: packages/api-capabilities/src/api_capabilities/_telemetry.py
from __future__ import annotations

from opentelemetry import trace
from opentelemetry.trace import Span, Status, StatusCode

from ._version import API_CAPABILITIES_VERSION

_TRACER_NAME = "meridian.api-capabilities"


def get_tracer() -> trace.Tracer:
    return trace.get_tracer(_TRACER_NAME, API_CAPABILITIES_VERSION)


def record_list_event(span: Span, *, count: int) -> None:
    """Attach a structured capabilities.list.invocation event to the active span."""
    span.add_event("capabilities.list.invocation", {"count": count})


def record_register_event(span: Span, *, namespace: str, capability_count: int) -> None:
    """Attach a structured capabilities.register.invocation event to the active span."""
    span.add_event(
        "capabilities.register.invocation",
        {"namespace": namespace, "capability_count": capability_count},
    )


def record_failure(span: Span, error: Exception, *, operation: str) -> None:
    """Set span status to ERROR, add a capabilities.error event, and record the exception."""
    span.set_status(Status(StatusCode.ERROR, str(error)))
    span.add_event(
        "capabilities.error",
        {
            "operation": operation,
            "error.type": type(error).__name__,
            "error.message": str(error),
        },
    )
    span.record_exception(error)

File: apps/meridian-ui/src/chat/telemetry.ts
import { SpanStatusCode, trace } from "@opentelemetry/api";
import type { Span } from "@opentelemetry/api";
import type { AuditLog } from "../workspace/audit.js";
import type { AuditLogEntry } from "../workspace/types.js";

const TRACER_NAME = "meridian.chat_composer";
const TRACER_VERSION = "0.1.0";

export function getTracer() {
  return trace.getTracer(TRACER_NAME, TRACER_VERSION);
}

export interface ComposerInvocationEvent {
  readonly sessionId: string;
  readonly timestamp: string;
  readonly channelId?: string;
  readonly agentId?: string;
  readonly contentLength: number;
}

export function recordComposerInvocationEvent(span: Span, event: ComposerInvocationEvent): void {
  span.addEvent("chat_composer.invocation", {
    "session.id": event.sessionId,
    "event.timestamp": event.timestamp,
    "message.content_length": event.contentLength,
    ...(event.channelId ? { "channel.id": event.channelId } : {}),
    ...(event.agentId ? { "agent.id": event.agentId } : {}),
  });
}

export function recordComposerFailure(
  span: Span,
  err: unknown,
  auditLog: AuditLog,
  context: { sessionId: string; operation: string },
): void {
  const message = err instanceof Error ? err.message : String(err);
  span.setStatus({ code: SpanStatusCode.ERROR, message });
  if (err instanceof Error) span.recordException(err);

  const entry: AuditLogEntry = {
    level: "error",
    event: `chat_composer.${context.operation}.failed`,
    sessionId: context.sessionId,
    timestamp: new Date().toISOString(),
    detail: { message },
  };
  auditLog.write(entry);
}
